From d81d49e746e15dad35de39dc0ace0cedfba7d2f7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?St=C3=A9phane=20Graber?= <stgraber@stgraber.org>
Date: Tue, 24 Mar 2026 16:40:16 -0400
Subject: [PATCH] incus/remote: Actually validate the token code in `incus
 webui`
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Fix the logic to validate the token both on initial use and through cookie.
Also add an Origin check for good measure.

This addresses CVE-2026-33898

Reported-by: https://7asecurity.com
Signed-off-by: Stéphane Graber <stgraber@stgraber.org>
---
 cmd/incus/remote_unix.go | 21 ++++++++++++++++++++-
 cmd/incus/webui_unix.go  |  1 +
 2 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/cmd/incus/remote_unix.go b/cmd/incus/remote_unix.go
index a328f941757..43aa14df1d8 100644
--- a/cmd/incus/remote_unix.go
+++ b/cmd/incus/remote_unix.go
@@ -214,6 +214,7 @@ func (t remoteProxyTransport) RoundTrip(r *http.Request) (*http.Response, error)
 type remoteProxyHandler struct {
 	s         incus.InstanceServer
 	transport http.RoundTripper
+	url       string
 
 	mu           *sync.RWMutex
 	connections  *uint64
@@ -238,6 +239,12 @@ func (h remoteProxyHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	*h.connections++
 	h.mu.Unlock()
 
+	// Don't allow cross-origin requests.
+	origin := r.Header.Get("Origin")
+	if origin != "" && origin != h.url {
+		return
+	}
+
 	// Basic auth.
 	if h.token != "" {
 		// Parse query URL.
@@ -248,6 +255,8 @@ func (h remoteProxyHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 		token := values.Get("auth_token")
 		if token != "" {
+			// If a token was passed through the URL, persist it as a cookie.
+
 			tokenCookie := http.Cookie{
 				Name:     "auth_token",
 				Value:    token,
@@ -259,11 +268,21 @@ func (h remoteProxyHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 			http.SetCookie(w, &tokenCookie)
 		} else {
+			// If not, attempt to pull it from the cookie.
 			cookie, err := r.Cookie("auth_token")
-			if err != nil || cookie.Value != h.token {
+			if err != nil {
+				// Fail authentication if no cookie can be found.
 				w.WriteHeader(http.StatusUnauthorized)
 				return
 			}
+
+			token = cookie.Value
+		}
+
+		// Check the user token against the expected value.
+		if token != h.token {
+			w.WriteHeader(http.StatusUnauthorized)
+			return
 		}
 	}
 
diff --git a/cmd/incus/webui_unix.go b/cmd/incus/webui_unix.go
index 44e0b967b81..9e20b035227 100644
--- a/cmd/incus/webui_unix.go
+++ b/cmd/incus/webui_unix.go
@@ -98,6 +98,7 @@ func (c *cmdWebui) Run(cmd *cobra.Command, args []string) error {
 		transactions: &transactions,
 
 		token: token,
+		url:   "http://" + server.Addr().String(),
 	}
 
 	// Print address.
